go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -920,30 +639,62 @@ func (bnc *BaseNetworkController) WatchPods() error {
 	return err
 }
 
-func calculateStaticIPs(podDesc string, ips []string) ([]*net.IPNet, error) {
-	var staticIPs []*net.IPNet
-	klog.V(5).Infof("Pod %s requested static IPs: %s", podDesc, strings.Join(ips, ";"))
-	for _, ip := range ips {
-		ipAddr, ipNet, err := net.ParseCIDR(ip)
+// allocateIPs and update the corresponding pod annotation
+func (bnc *BaseNetworkController) allocateIPs(pod *kapi.Pod, lsp *nbdb.LogicalSwitchPort, nadName string, network *nadapi.NetworkSelectionElement) (*util.PodAnnotation, bool, error) {
+	switchName, err := bnc.getExpectedSwitchName(pod)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if network == nil {
+		network = &nadapi.NetworkSelectionElement{}
+	}
+
+	var reallocate bool
+	if lsp != nil && len(network.IPRequest) == 0 {
+		mac, ips, err := bnc.getPortAddresses(switchName, lsp)
 		if err != nil {
-			return nil, fmt.Errorf("failed to parse IP %s requested in annotation for pod %s: Error %v",
-				ip, podDesc, err)
+			return nil, false, fmt.Errorf("failed to get pod addresses for pod %s/%s/%s on node %s, err: %v",
+				nadName, pod.Namespace, pod.Name, switchName, err)
 		}
-		ipNet.IP = ipAddr
-		staticIPs = append(staticIPs, ipNet)
+		network.MacRequest = mac.String()
+		network.IPRequest = util.StringSlice(ips)
+		reallocate = true
+
+		klog.V(5).Infof("Will attempt to use LSP IP addresses %v and mac %s for pod %s/%s/%s",
+			network.IPRequest, network.MacRequest, nadName, pod.Namespace, pod.Name)
 	}
 
-	return staticIPs, nil
-}
+	var ipAllocator *subnetipallocator.IPAllocator
+	if bnc.doesNetworkRequireIPAM() {
+		ipAllocator = bnc.lsManager.ForSwitch(switchName)
+	}
+
+	updatedPod, podAnnotation, err := util.AllocatePodAnnotation(
```

Comment:
seems like we are passing a lot of arguments to this function with bnc properties. Why not make it a receiver?

Reply:
I can make a new struct for it that we initialize with some of the args and can act as receiver. WDYT?

Reply:
I was thinking we can just make some new interface that bnc inherits and ovn_cluster_manager can also use. I guess  maybe it is similar to the interface you have in util/pod_annotation.go. Then we can make struct that implements that interface, and make these functions be receivers to that.

Reply:
That seem to it would be inheriting just to share code. There should be a clear composition relationship for that. We can just have an utility struct and use it where needed.